### PR TITLE
added users table, entity, dtos, service, repository, and password an…

### PR DESCRIPTION
**Pull request by HAIOVYI**
…notation for matching

### CODE REVIEW

File: src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+@Component
+public class PasswordMatchesValidator implements ConstraintValidator<FieldMatch, Object> {
+    private String firstFieldName;
+    private String secondFieldName;
+
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        firstFieldName = constraintAnnotation.first();
+        secondFieldName = constraintAnnotation.second();
+    }
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java:17]: What is the reason for secondFieldName, firstFieldName? This fields is not used

File: src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    @Override
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        UserRegistrationRequestDto userDto = (UserRegistrationRequestDto) value;
+        String password = userDto.getPassword();
+        String confirmPassword = userDto.getConfirmPassword();
+
+        return password != null && password.equals(confirmPassword);
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java:25]: Please use Objects.equals

File: src/main/java/mate/academy/mybookshop/dto/UserRegistrationRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+import mate.academy.mybookshop.annotation.FieldMatch;
+
+@FieldMatch.List({
+        @FieldMatch(first = "password",
+                second = "confirmPassword",
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/mybookshop/dto/UserRegistrationRequestDto.java:11]: This parameters are not used

File: src/main/java/mate/academy/mybookshop/service/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    @Override
+    public UserResponseDto registerUser(UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.existsByEmail(request.getEmail())) {
+            throw new RegistrationException("Email already exists");
+        }
+
+        if (!passwordMatchesValidator.isValid(request, null)) {
+            throw new RegistrationException("Passwords do not match");
+        }
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/mybookshop/service/UserServiceImpl.java:29]: This should happen automatically by spring

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 5]
```diff
@@ -1,4 +1,6 @@
 databaseChangeLog:
   - include:
       file: db/changelog/changes/01-create-books-table.yaml
+  - include:
+      file: db/changelog/changes/02-create-users-table.yaml
```
Comment (andrii-hoienko) [inline src/main/resources/db/changelog/db.changelog-master.yaml:5]:
```suggestion
      file: db/changelog/changes/02-create-users-table.yaml
      
```

File: src/main/java/mate/academy/mybookshop/mapper/UserMapper.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+import mate.academy.mybookshop.config.MapperConfig;
+import mate.academy.mybookshop.dto.UserRegistrationRequestDto;
+import mate.academy.mybookshop.dto.UserResponseDto;
+import mate.academy.mybookshop.entity.UserEntity;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface UserMapper {
+    UserResponseDto userToUserResponseDto(UserEntity userEntity);
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/mybookshop/mapper/UserMapper.java:11]:
```suggestion
    UserResponseDto toResponseDto(UserEntity userEntity);
```

File: src/main/java/mate/academy/mybookshop/mapper/UserMapper.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+    UserEntity userRegistrationRequestDtoToUserEntity(
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/mybookshop/mapper/UserMapper.java:13]:
```suggestion
    UserEntity toEntity(
```

File: pom.xml [hunk 8]
```diff
@@ -82,6 +82,16 @@
...
+        <dependency>
+            <groupId>javax.validation</groupId>
+            <artifactId>validation-api</artifactId>
+            <version>1.0.0.GA</version>
+        </dependency>
+        <dependency>
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-validator</artifactId>
+            <version>4.0.2.GA</version>
+        </dependency>
```
Comment (andrii-hoienko) [inline pom.xml:94]: Please use properties for versions

File: pom.xml [hunk 9]
```diff
@@ -82,6 +82,16 @@
             <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
             <version>2.1.0</version>
         </dependency>
+        <dependency>
+            <groupId>javax.validation</groupId>
+            <artifactId>validation-api</artifactId>
+            <version>1.0.0.GA</version>
```
Comment (andrii-hoienko) [inline pom.xml:88]: I have a feeling that this dependency is wrong, check this one https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-validation

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job, but some improvements required

File: src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java (hunk 2 above)
Comment (HAIOVYI) [reply src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java:25]: thx

File: src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java (hunk 1 above)
Comment (HAIOVYI) [reply src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java:17]: thx

File: pom.xml (hunk 8 above)
Comment (HAIOVYI) [reply pom.xml:94]: thx

File: pom.xml (hunk 9 above)
Comment (HAIOVYI) [reply pom.xml:88]: thx

File: src/main/java/mate/academy/mybookshop/service/UserServiceImpl.java (hunk 4 above)
Comment (HAIOVYI) [reply src/main/java/mate/academy/mybookshop/service/UserServiceImpl.java:29]: fixed

File: src/main/java/mate/academy/mybookshop/dto/UserRegistrationRequestDto.java (hunk 3 above)
Comment (HAIOVYI) [reply src/main/java/mate/academy/mybookshop/dto/UserRegistrationRequestDto.java:11]: fixed

File: src/main/java/mate/academy/mybookshop/mapper/UserMapper.java (hunk 6 above)
Comment (HAIOVYI) [reply src/main/java/mate/academy/mybookshop/mapper/UserMapper.java:11]: thx

File: src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import jakarta.validation.ConstraintValidatorContext;
+import java.util.Objects;
+import mate.academy.mybookshop.dto.UserRegistrationRequestDto;
+
+public class PasswordMatchesValidator implements ConstraintValidator<PasswordMatches, Object> {
+
+    @Override
+    public void initialize(PasswordMatches constraintAnnotation) {
+        //ignored
+    }
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java:13]: Is it necessary to override it?

File: src/main/java/mate/academy/mybookshop/service/UserService.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.mybookshop.service;
+
+import mate.academy.mybookshop.dto.UserRegistrationRequestDto;
+import mate.academy.mybookshop.dto.UserResponseDto;
+import mate.academy.mybookshop.exception.RegistrationException;
+import org.springframework.stereotype.Service;
+
+@Service
+public interface UserService {
+    UserResponseDto registerUser(UserRegistrationRequestDto request) throws RegistrationException;
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/mybookshop/service/UserService.java:10]:
```suggestion
    UserResponseDto registerUser(UserRegistrationRequestDto request);
```

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Left minor comments and question

File: src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java (hunk 10 above)
Comment (HAIOVYI) [reply src/main/java/mate/academy/mybookshop/annotation/PasswordMatchesValidator.java:13]: nope, fixed

File: src/main/java/mate/academy/mybookshop/service/UserService.java (hunk 11 above)
Comment (HAIOVYI) [reply src/main/java/mate/academy/mybookshop/service/UserService.java:10]: fixed

**Review by andrii-hoienko — APPROVED** [review]
Good job!

**Pull request merged**